docs: dockertest covid article (#85)
ory/web: src/pages/blog/ory-vs-covid.md

---

> path: '/ory-vs-covid/'

Reviewer: ```suggestion
path: '/dockertest-gaen-google-apple-exposure-notification-covid-19/'
```

title: |

>   How Ory is helping keep Covid in check

Reviewer: ```suggestion
  Keeping Covid-19 in check with ORY Dockertest
```

teaser: |
  This article gives a short introduction to Dockertest and how Google is using it against Covid19,
  also an overview about the open-source virus response in general. 

seo:
  title: |

>     Open source software critical for global virus response

Reviewer: ```suggestion
    Write tests like Google and Apple Exposure Notifications using Dockertest
```

---
path: '/ory-vs-covid/'

title: |
  How Ory is helping keep Covid in check

teaser: |
  This article gives a short introduction to Dockertest and how Google is using it against Covid19,
  also an overview about the open-source virus response in general. 

seo:
  title: |
    Open source software critical for global virus response
  keywords: |
    docker, test, database testing, cloud development, auth, authentication, open source, Covid19, pandemic

publishedAt: '2020-12-02'
author: vinckr

overline: |
  Dockertest - Automated Database Testing in Docker   
---

> ### What is Ory Dockertest?

Reviewer: Do not start posts directly with a headline or this happens:

<img width="1207" alt="Bildschirmfoto 2020-11-27 um 15 08 18" src="https://user-images.githubusercontent.com/3372410/100457499-6e76cc00-30c2-11eb-83c4-5e4471bbcf7d.png">

Using an API abstraction simplifies the process. But for small changes to the table structure you need to work on the mocks or deal with other ramifications in the API.  
You have the problem that the mock is not a real database. You also have to rewrite the mock database and unit tests, when you make changes to your configuration.   
This is where Docker and Ory Dockertest can help you create real databases for your tests. You dont have to set up a whole database infrastructure.  

[Dockertest](https://github.com/ory/Dockertest) automates your interactions with the Docker API. This lets you test your services database integration in a simple and straightforward way. 
You configure your specific test setup, create the databases you want to test, ping to ensure connection and after testing close and dispose of the database again.  
This process becomes 

In [Dockertest](https://github.com/ory/Dockertest) a pool represents a connection to the Docker API and is used to handle docker images and resources represent the docker containers.
You can find examples for all common databases on [github.com/ory/Dockertest](https://github.com/ory/Dockertest). 

> ### How Dockertest is helping keeping Covid in check

Reviewer: This post has no H2 (`##`). Make sure to keep the structure of h2,h3,h4,h5, ...

https://yoast.com/how-to-use-headings-on-your-site/

---
published: true
path: '/ory-vs-covid/'
title: >
  How Ory is helping keep Covid in check

metaDescription: >
  Dockertest makes database testing so easy, even Google is jealous. 

metaTitle: >
  Open source software critical for global virus response

publishedAt: '2020-11-27'
author: 'Vincent Kraus'
overline: >
  Dockertest - easy go tests with docker

category: Article
subtitle: >

>  How Open-source software and Ory can be part of a greater effort against the virus

Reviewer 2: This was removed recently, adapt according to the other PRs :wink:

Goverments worldwide use it to provide their citizens with a means of tracing infections or rather notifiying them of exposure to the virus. 
Since most people either have an Apple or Android device this approach covers a broad chunk of the population. 

Now what does this have to do with [Dockertest](https://github.com/ory/Dockertest)?

Google's [open source reference implementation](https://github.com/google/exposure-notifications-server) of an notification server uses [Dockertest](https://github.com/ory/Dockertest) for [database testing](https://github.com/google/exposure-notifications-server/blob/c35c97708dec0a8a3e2ac54f89f9033be5da22a6/internal/database/database_util.go). 
It is amazing to see a tool that was written to ease the development process at Ory is used by big tech organisations in the most impactful health operation in modern times. It shows what a reliable and essential tool [Dockertest](https://github.com/ory/Dockertest) has become over the years.

Here you can see the full code of the implementation: 

> ```

Reviewer 2: ```suggestion
```go
```

Reviewer 2: Or if the code is copied from GH, use https://github.com/ory/web/pull/75/files#diff-9cdb5dd0b1f36dc0efaf706b88cb1744be3b404f4095e8c4e384b51cd046ba94R128-R132

> 
>  NewTestDatabaseWithConfig creates a new database suitable for use in testing.

Reviewer 2: ```suggestion
// NewTestDatabaseWithConfig creates a new database suitable for use in testing.
```

	// but there's no point in trying immediately.
	time.Sleep(1 * time.Second)

	b, err := retry.NewFibonacci(500 * time.Millisecond)
	if err != nil {
		tb.Fatalf("failed to configure backoff: %v", err)
	}
	b = retry.WithMaxRetries(10, b)
	b = retry.WithCappedDuration(10*time.Second, b)

	// Establish a connection to the database. Use a Fibonacci backoff instead of
	// exponential so wait times scale appropriately.
	var dbpool *pgxpool.Pool
	if err := retry.Do(ctx, b, func(ctx context.Context) error {
		var err error
		dbpool, err = pgxpool.Connect(ctx, connURL.String())
		if err != nil {
			return retry.RetryableError(err)
		}
		return nil
	}); err != nil {
		tb.Fatalf("failed to start postgres: %s", err)
	}

	// Run the migrations.
	if err := dbMigrate(connURL.String()); err != nil {
		tb.Fatalf("failed to migrate database: %s", err)
	}

	// Create the db instance.
	db := &DB{Pool: dbpool}

	// Close db when done.
	tb.Cleanup(func() {
		db.Close(context.Background())
	})

	return db, &Config{
		Name:     dbname,
		User:     username,
		Host:     container.GetBoundIP("5432/tcp"),
		Port:     container.GetPort("5432/tcp"),
		SSLMode:  "disable",
		Password: password,
	}
}

> 

Reviewer 2: ```suggestion
```
